File: streaming-s3-example-java/src/main/java/nl/vintik/streaming/java/ParseResult.java
package nl.vintik.streaming.java;

/**
 * Outcome of parsing the API Gateway proxy event from the input stream.
 *
 * <p>Drives the handler's status mapping deterministically via a pattern-matching
 * {@code switch}: {@link Parsed} carries the domain request; {@link ParseError} maps to
 * HTTP 400 (request could not be parsed).
 */
public sealed interface ParseResult permits ParseResult.Parsed, ParseResult.ParseError {

    /** The proxy event was decoded successfully into a {@link StreamRequest}. */
    record Parsed(StreamRequest request) implements ParseResult {}

    /** The proxy event could not be decoded (malformed, empty, or truncated JSON). */
    record ParseError() implements ParseResult {}
}

File: streaming-s3-example/src/main/kotlin/nl/vintik/lambda/streaming/ParseResult.kt
package nl.vintik.lambda.streaming

/**
 * Outcome of parsing the API Gateway proxy event from the input stream.
 *
 * Drives the handler's status mapping deterministically: [Parsed] carries the
 * domain request; [ParseError] maps to HTTP 400 (request could not be parsed).
 */
sealed interface ParseResult {
    data class Parsed(val request: StreamRequest) : ParseResult
    data object ParseError : ParseResult
}
